src/DataFrame/Operations/Typing.hs

```diff
@@ -10,110 +10,183 @@ import qualified Data.Text as T
-            notAllParsed = V.any isLeft safeVector
-         in if safeRead && notAllParsed
-                then
-                    if V.all (either isNullish (const False)) (V.filter isLeft safeVector)
-                        then OptionalColumn (V.map (either (const Nothing) Just) safeVector)
-                        else BoxedColumn safeVector
-                else
-                    UnboxedColumn
-                        (VU.generate (V.length c) (either undefined id . (safeVector V.!)))
-    | V.any isJust (V.map (parseTimeOpt dateFormat) examples) =
-        let
-            -- failed parse should be Either, nullish should be Maybe
-            emptyToNothing' v = if isNullish v then Left v else Right v
-            parseTimeEither v = case parseTimeOpt dateFormat v of
-                Just v' -> Right v'
-                Nothing -> Left v
-            safeVector = V.map ((=<<) parseTimeEither . emptyToNothing') c
-            toMaybe (Left _) = Nothing
-            toMaybe (Right value) = Just value
-            lefts = V.filter isLeft safeVector
-            onlyNulls = (not (V.null lefts) && V.all (isNullish . fromLeft "non-null") lefts)
-         in
-            if safeRead
-                then
-                    if onlyNulls
-                        then OptionalColumn (V.map toMaybe safeVector)
-                        else
-                            if V.any isLeft safeVector
-                                then BoxedColumn safeVector
-                                else BoxedColumn (V.map (unsafeParseTime dateFormat) c)
-                else BoxedColumn (V.map (unsafeParseTime dateFormat) c)
-    | otherwise =
-        let
-            safeVector = V.map emptyToNothing c
-            hasNulls = V.any isNullish c
-         in
-            if safeRead && hasNulls then OptionalColumn safeVector else BoxedColumn c
+parseFromExamples :: Int -> Bool -> DateFormat -> V.Vector T.Text -> Column
+parseFromExamples n safeRead dateFormat cols =
+    let
+        converter = if safeRead then convertNullish else convertOnlyEmpty
+        examples = V.map converter (V.take n cols)
+        asMaybeText = V.map converter cols
+     in
+        case makeParsingAssumption dateFormat examples of
+            IntAssumption -> handleIntAssumption asMaybeText
+            DoubleAssumption -> handleDoubleAssumption asMaybeText
+            TextAssumption -> handleTextAssumption asMaybeText
+            DateAssumption -> handleDateAssumption dateFormat asMaybeText
+            NoAssumption -> handleNoAssumption dateFormat asMaybeText
+
+handleIntAssumption :: V.Vector (Maybe T.Text) -> Column
+handleIntAssumption asMaybeText
+    | not (hasNullValues asMaybeInt) = UnboxedColumn $ prepForUnboxed asMaybeInt
+    | not (hasNullValues asMaybeDouble) =
+        UnboxedColumn $ prepForUnboxed asMaybeDouble
+    | parsableAsInt = OptionalColumn asMaybeInt
+    | parsableAsDouble = OptionalColumn asMaybeDouble
```

Comment:
We have a `fromVector` function in the column module that automatically puts things in the right column type (boxed, optional etc).